log: Mitigate disk filling attacks by rate limiting LogPrintf

dergoegge opened the pull request:
This picks up #19995 in an attempt to solve #21559.

The goal of this PR is to mitigate disk filling attacks by temporarily rate limiting `LogPrintf(…)`.

A disk fill attack is an attack where an untrusted party (such as a peer) is able to cheaply make your node log to disk excessively. The excessive logging may fill your disk and thus make your node crash either cleanly (best case: if disk fill rate is relatively slow) or uncleanly (worst case: if disk fill rate is relatively fast).

## Approach

The hourly logging quota is set **per source location**. Every single source location (say net_processing.cpp:418) gets a quota of 1 MB of logging per hour. Logging is only dropped from source locations that exceed the quota.

- Only logging to disk is rate limited. Logging to console is not rate limited.
- Only `LogPrintf(…)` is rate limited. `LogPrint(category, …)` (`-debug`) is not rate limited.
- `UpdateTip: new best=[…]` is logged using `LogPrintfWithoutRateLimiting(…)` to avoid rate limiting. High log volume is expected for that source location during IBD.
- When logging is restarted a tally of how many messages/bytes were dropped is printed.
- Log rate limiting can be disabled with the `-ratelimitlogging` config option.
- All logs will be prefixed with `[*]` if there is at least one source location that is currently being suppressed.

### Alternatives

In a recent [PR Review Club meeting](https://bitcoincore.reviews/21603) we discussed this PR and evaluated alternative approaches.

<details>
<summary>Global rate limiting </summary>

#### Global rate limiting

There was some discussion around an alternative approach which would be to globally rate limit `LogPrintf(…)`.  This approach was implemented in a competing PR #21706. The main point [1] in favor of global rate limiting is that it could be confusing to have incomplete logs when logs from one source location are dropped but not from others. The main point against global rate limiting is that it opens another attack vector where an attacker could trigger rate limiting and then execute a 2. attack which would then not be document in the logs at all. With regard to the attack vector on the global approach and the overall reviews the two approaches have gotten I have chosen to continue with the approach in this PR. 
(To address the concern of [1] i have chosen to prefix all logs with `[*]` if there is at least one source location that is currently being suppressed.)
</details>

<details>
<summary>logrotate</summary>

#### logrotate

LarryRuane brought up [logrotate](https://linux.die.net/man/8/logrotate) which can be used on linux to monitor log files and perform actions (delete, compress, send somewhere, etc.) on them when a certain size or point in time is reached. `logrotate` could be used to rate limit log but since we would like log rate limiting to be enabled by default it is not the best solution. Also on windows an alternative would have to be used.

</details>

## Testing

I have a written a unit test which is contained in this PR. (Not the greatest test, it currently does not work on windows thanks to `\r\n`. If someone has an idea how to improve the test i would appreciate it)

Further more I made a RPC [here](https://github.com/dergoegge/bitcoin/commit/8e6d15d6be06834d1f634f4a54ea646be8bc3491) that can log excessive amounts of "a"s from different locations which i have found useful during my own testing.
`bitcoin-cli excessivelog <location (1-5)> <num_bytes>`

⚠️One thing to note with that rpc is that the rate limiting logic still prints the last message that triggered the limiting to disk, so something like `bitcoin-cli excessivelog 1 536870912` would still log ~512MiB to disk. Logging to console is also never suppressed (unless -printtoconsole=0)  ⚠️

A simple example to use the rpc:
```bash
bitcoin-cli -regtest setmocktime 1
bitcoin-cli -regtest excessivelog 1 1048500 # log just under 1MiB
bitcoin-cli -regtest excessivelog 1 100 # this should get the total amount logged above 1MiB
                                        # and the rate limiting logic should kick in
bitcoin-cli -regtest setmocktime 3602 # let 1 hour pass (only works in regtest)
bitcoin-cli -regtest excessivelog 1 100 # this should trigger logging to resume
```

dergoegge commented:
In response to @jnewbery's [comment](https://github.com/bitcoin/bitcoin/pull/19995#pullrequestreview-505540213)

> I think I'd prefer log-shedding to be done globally, with a special log at the end of a log suppression period that tallies up which log locations have been suppressed.

I disagree on the global schedding, i think the source location based schedding is over all the better approach.
I agree with your point that it could be confusing to drop some logs in the context of user issues but schedding should only kick in if there is an attack or something is horribly wrong, so i would think that almost all logs provided by users would be complete. We can also always tell if rate limiting is taking place since we log when it starts/stops and in those cases a partial log is better then no log.

(This is obviously mostly an opinion and I am totally willing to be convinced of global log-schedding if enough reviewers favor it)

jnewbery commented:
> schedding should only kick in if there is an attack or something is horribly wrong, so i would think that almost all logs provided by users would be complete.

Right, and those cases are exactly where we need the best possible logs.

> We can also always tell if rate limiting is taking place since we log when it starts/stops and in those cases a partial log is better then no log.

I guarantee you that not everyone will know this and most people will miss a single log line saying "rate-limiting log x".

I can only speak from my own experience of supporting network equipment for several years - partial logs are often worse than no logs at all since they're so misleading.

practicalswift commented:
Strongest possible Concept ACK: this mitigation will kill an entire bug class (the "disk fill via logging" bug class). 

Thanks for taking up this work @dergoegge.

Will test and review.

practicalswift commented:
@jnewbery 

> I think I'd prefer log-shedding to be done globally, with a special log at the end of a log suppression period that tallies up which log locations have been suppressed.

Clarifying questions to fully understand what is suggested:

Let's assume that log location A is a misplaced `LogPrintf` in a code path that the attacker can make us take.

In the case of an attacker using log location A to fill our disk by making us log from repeatedly, is the suggestion then that _all_ log locations should be suppressed during the suppression period (instead of suppressing only the "attacker controlled" log location A)?

To make your suggestion clear: could you exemplify what the special log entry at the end of the log suppression period would look like? (A patch or even a separate PR would be even better, but an example log entry would probably clarify enough!)

adamjonas commented on src/test/logging_tests.cpp:
```diff
@@ -33,4 +32,109 @@ BOOST_AUTO_TEST_CASE(logging_timer)
+template <typename... Args>
+void LogFromFixedLocation(const char* fmt, const Args&... args)
+{
+    LogPrintf(fmt, args...);
```
this line is tripping up the linter:
```
src/test/logging_tests.cpp: Expected 0 argument(s) after format string but found 1 argument(s): LogPrintf(fmt, args...)
```

adamjonas commented on src/test/logging_tests.cpp:
```diff
@@ -33,4 +32,108 @@ BOOST_AUTO_TEST_CASE(logging_timer)
+        LogFromFixedLocation(s1023);
+    }
+    GetLogFileSize(curr_log_file_size);
+    BOOST_CHECK(curr_log_file_size - prev_log_file_size == 1024 * 1024);
```
Both windows CI builds are failing on this line. See [cirrus](https://cirrus-ci.com/task/6258137339527168?logs=ci#L642).

practicalswift commented:
@dergoegge 

Thanks for your work on this PR!

Would you be willing to implement also @jnewbery's suggestions as a separate PR?

I tried to summarise his suggestion in https://github.com/bitcoin/bitcoin/pull/21603#issuecomment-813933612. I'm not certain I got it right though - @jnewbery, feel free to chime in :)

I think it makes sense to do it as separate PR since it deviates in important ways from the original suggestion. By having two separate PRs we would see which of the two approaches to address this type of attack that have consensus support.

Personally I think I could live with both approaches: as long as we kill the disk-fill-via-logging bug class sooner rather than later I'm happy :)

dergoegge commented:
> Would you be willing to implement also jnewbery's suggestions as a separate PR?

@practicalswift Sure, thats a good idea :)

rebroad commented:
concept ACK (not that I'm sure if my opinion is relevant) and I think this makes more sense than the global alternative.

practicalswift commented:
@dergoegge Thanks for working on this. Would you mind rebasing? I would like to review the updated version :)

dergoegge commented:
@practicalswift Rebased :)

For anyone reviewing this: I made a rpc [here](https://github.com/dergoegge/bitcoin/commit/8e6d15d6be06834d1f634f4a54ea646be8bc3491) that can log excessive amounts of "a"s from different locations which might be useful for testing.
`bitcoin-cli excessivelog <location (1-5)> <num_bytes>`

⚠️One thing to note with that rpc is that the rate limiting logic still prints the last message that triggered the limiting to disk, so something like `bitcoin-cli excessivelog 1 536870912` would still log ~512MiB to disk. Logging to console is also never suppressed (unless -printtoconsole=0)  ⚠️

A simple example to use the rpc:
```bash
bitcoin-cli excessivelog 1 1048500 # log just under 1MiB
bitcoin-cli excessivelog 1 100 # this should get the total amount logged above 1MiB and the rate limiting logic should kick in
```

Emzy commented:
concept ACK 
I think this PR makes more sense than the global alternative, because global rate limiting would open another attack vector.
An attacker could trigger the global limit in one category and conceal another attack from the log.

dergoegge commented:
I moved the rate limiting logic into its own method.
[log_ratelimiting.0](https://github.com/dergoegge/bitcoin/tree/log_ratelimiting.0) -> [log_ratelimiting.1](https://github.com/dergoegge/bitcoin/tree/log_ratelimiting.1)

<details>
<summary>git range-diff master log_ratelimiting.0 log_ratelimiting.1</summary>

```
1:  76c59a1cc ! 1:  a3a2cfd64 log: Mitigate disk filling attacks by rate limiting LogPrintf
    @@ Metadata
      ## Commit message ##
         log: Mitigate disk filling attacks by rate limiting LogPrintf
     
    +    Co-authored-by: dergoegge <n.goeggi@gmail.com>
    +
      ## src/init/common.cpp ##
     @@ src/init/common.cpp: void AddLoggingArgs(ArgsManager& argsman)
          argsman.AddArg("-logtimemicros", strprintf("Add microsecond precision to debug timestamps (default: %u)", DEFAULT_LOGTIMEMICROS), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
    @@ src/logging.cpp: namespace BCLog {
     -void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& logging_function, const std::string& source_file, const int source_line)
     +static constexpr uint64_t HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION{1024 * 1024};
     +
    ++bool BCLog::Logger::RateLimit(std::string& str, const std::string& logging_function, const SourceLocation& source_location)
    ++{
    ++    if (!m_rate_limiting) {
    ++        // Rate limiting is disabled.
    ++        return false;
    ++    }
    ++
    ++    const std::chrono::seconds now = GetTime<std::chrono::seconds>();
    ++    QuotaUsage& quota_usage = m_quota_usage_per_source_location[source_location];
    ++    // Is the quota exceeded before this log call?
    ++    bool quota_exceeded_before = quota_usage.m_bytes_logged > HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION;
    ++
    ++    bool dont_skip = false;
    ++    // Every hour the quota usage for a source location is reset.
    ++    if ((now - quota_usage.m_last_reset) > std::chrono::hours{1}) {
    ++        // Should logging to disk continue to be disabled?
    ++        bool quota_still_exceeded = quota_usage.m_bytes_dropped > HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION;
    ++
    ++        if (quota_still_exceeded) {
    ++            str = LogTimestampStr(strprintf(
    ++                "Not restarting logging from %s:%d (%s): "
    ++                "because %d messages (%d MiB) were dropped during the last hour which still exceeds the limit of %d MiB.\n",
    ++                source_location.first, source_location.second, logging_function, quota_usage.m_messages_dropped,
    ++                quota_usage.m_bytes_dropped / (1024 * 1024), HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION / (1024 * 1024)));
    ++        } else if (quota_exceeded_before) {
    ++            str = LogTimestampStr(strprintf(
    ++                "Restarting logging from %s:%d (%s): "
    ++                "%d messages (%d MiB) were dropped during the last hour.\n"
    ++                "%s",
    ++                source_location.first, source_location.second, logging_function, quota_usage.m_messages_dropped,
    ++                quota_usage.m_bytes_dropped / (1024 * 1024), str));
    ++            --m_rate_limited_locations;
    ++        }
    ++
    ++        // Dont skip the reset logs.
    ++        dont_skip = quota_exceeded_before || quota_still_exceeded;
    ++
    ++        // Logging to disk is only re-enabled if the number of dropped bytes did not exceed the limit.
    ++        if (!quota_still_exceeded) quota_usage.m_bytes_logged = 0;
    ++        quota_usage.m_messages_dropped = 0;
    ++        quota_usage.m_bytes_dropped = 0;
    ++        quota_usage.m_last_reset = now;
    ++    }
    ++
    ++    if (!quota_exceeded_before) quota_usage.m_bytes_logged += str.size();
    ++
    ++    bool quota_exceeded_after = quota_usage.m_bytes_logged > HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION;
    ++    if (!quota_exceeded_after) {
    ++        // The limits were not exceeded and the message should not be dropped.
    ++        return false;
    ++    }
    ++
    ++    if (!quota_exceeded_before) {
    ++        str = LogTimestampStr(strprintf(
    ++            "Excessive logging detected from %s:%d (%s): "
    ++            ">%d MiB logged during the last hour. "
    ++            "Suppressing logging to disk from this source location for up to one hour. "
    ++            "Console logging unaffected. Last log entry: %s",
    ++            source_location.first, source_location.second, logging_function,
    ++            HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION / (1024 * 1024), str));
    ++        ++m_rate_limited_locations;
    ++    } else if (!dont_skip) {
    ++        // The log message should be dropped.
    ++        quota_usage.m_messages_dropped++;
    ++        quota_usage.m_bytes_dropped += str.size();
    ++        return true;
    ++    }
    ++
    ++    return false;
    ++}
    ++
     +void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& logging_function, const SourceLocation& source_location, const bool skip_disk_usage_rate_limiting)
      {
          StdLockGuard scoped_lock(m_cs);
    @@ src/logging.cpp: void BCLog::Logger::LogPrintStr(const std::string& str, const s
      
     +    // Rate limit logging to disk to avoid disk filling attacks.
     +    bool skip_writing_to_disk_due_to_rate_limiting{false};
    -+    if (!skip_disk_usage_rate_limiting && m_rate_limiting) {
    -+        const std::chrono::seconds now = GetTime<std::chrono::seconds>();
    -+        QuotaUsage& quota_usage = m_quota_usage_per_source_location[source_location];
    -+        // Is the quota exceeded before this log call?
    -+        bool quota_exceeded_before = quota_usage.m_bytes_logged > HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION;
    -+
    -+        bool dont_skip = false;
    -+        // Every hour the quota usage for a source location is reset.
    -+        if ((now - quota_usage.m_last_reset) > std::chrono::hours{1}) {
    -+            // Should logging to disk continue to be disabled?
    -+            bool quota_still_exceeded = quota_usage.m_bytes_dropped > HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION;
    -+
    -+            if (quota_still_exceeded) {
    -+                str_prefixed = LogTimestampStr(strprintf("Not restarting logging from %s:%d (%s): because %d messages (%d MiB) were dropped during the last hour which still exceeds the limit of %d MiB.\n", source_location.first, source_location.second, logging_function, quota_usage.m_messages_dropped, quota_usage.m_bytes_dropped / (1024 * 1024), HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION / (1024 * 1024)));
    -+            } else if (quota_exceeded_before) {
    -+                str_prefixed = LogTimestampStr(strprintf("Restarting logging from %s:%d (%s): %d messages (%d MiB) were dropped during the last hour.\n", source_location.first, source_location.second, logging_function, quota_usage.m_messages_dropped, quota_usage.m_bytes_dropped / (1024 * 1024)));
    -+            }
    -+
    -+            // Dont skip the reset logs.
    -+            dont_skip = quota_exceeded_before || quota_still_exceeded;
    -+
    -+            // Logging to disk is only re-enabled if the number of dropped bytes did not exceed the limit.
    -+            if (!quota_still_exceeded) quota_usage.m_bytes_logged = 0;
    -+            quota_usage.m_messages_dropped = 0;
    -+            quota_usage.m_bytes_dropped = 0;
    -+            quota_usage.m_last_reset = now;
    -+        }
    ++    if (!skip_disk_usage_rate_limiting) {
    ++        skip_writing_to_disk_due_to_rate_limiting = RateLimit(str_prefixed, logging_function, source_location);
    ++    }
     +
    -+        if (!quota_exceeded_before) quota_usage.m_bytes_logged += str_prefixed.size();
    -+
    -+        bool quota_exceeded_after = quota_usage.m_bytes_logged > HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION;
    -+        if (quota_exceeded_after) {
    -+            if (!quota_exceeded_before) {
    -+                str_prefixed = LogTimestampStr(strprintf("Excessive logging detected from %s:%d (%s): >%d MiB logged during the last hour. Suppressing logging to disk from this source location for up to one hour. Console logging unaffected. Last log entry: %s", source_location.first, source_location.second, logging_function, HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION / (1024 * 1024), str_prefixed));
    -+            } else if (!dont_skip) {
    -+                skip_writing_to_disk_due_to_rate_limiting = !dont_skip;
    -+                quota_usage.m_messages_dropped++;
    -+                quota_usage.m_bytes_dropped += str_prefixed.size();
    -+            }
    -+        }
    ++    if (m_rate_limited_locations > 0) {
    ++        str_prefixed.insert(0, "[*] ");
     +    }
     +
          m_started_new_line = !str.empty() && str[str.size()-1] == '\n';
    @@ src/logging.h: namespace BCLog {
              std::list<std::string> m_msgs_before_open GUARDED_BY(m_cs);
              bool m_buffering GUARDED_BY(m_cs) = true; //!< Buffer messages before logging can be started.
     +        std::unordered_map<SourceLocation, QuotaUsage, SourceLocationHasher> m_quota_usage_per_source_location GUARDED_BY(m_cs);
    ++        std::atomic_int m_rate_limited_locations;
      
              /**
               * m_started_new_line is a state variable that will suppress printing of
    @@ src/logging.h: namespace BCLog {
              fs::path m_file_path;
              std::atomic<bool> m_reopen_file{false};
      
    ++        /** Evaluate if a log message should not be printed to disk.
    ++         *
    ++         * @param[in,out] The log message.
    ++         *                The rate limiting logic message may change the message
    ++         *                to signal the start or end of a supression period.
    ++         * @param[in]     The function from which the log message originated.
    ++         * @param[in]     The source location from which the log message originated.
    ++         */
    ++        bool RateLimit(std::string& str, const std::string& logging_function, const SourceLocation& source_location) EXCLUSIVE_LOCKS_REQUIRED(m_cs);
    ++
              /** Send a string to the log output */
     -        void LogPrintStr(const std::string& str, const std::string& logging_function, const std::string& source_file, const int source_line);
     +        void LogPrintStr(const std::string& str, const std::string& logging_function, const SourceLocation& source_location, const bool skip_disk_usage_rate_limiting);
2:  855d05e43 = 2:  5b46965aa test: Add logging test for rate limiting
```
</details>

I also decided to move forward with the approach in this PR and will close #21706. See the updated PR description for reasoning.

Thoragh commented:
Concept ACK

Maybe it would be good to set m_last_reset to the current time when it start limiting so it takes 1 hour until the next reset?

dergoegge commented:
> Maybe it would be good to set m_last_reset to the current time when it start limiting so it takes 1 hour until the next reset?

@Thoragh Thanks for the suggestion, Im gonna test that out. It might make sense to do but it does make the approach a bit more conservative because a source location might have to wait longer before it can log again (at least 1hour).

LarryRuane commented on src/logging.cpp:
```diff
@@ -254,6 +327,16 @@ void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& loggi
+    }
+
+    if (m_rate_limited_locations > 0) {
+        str_prefixed.insert(0, "[*] ");
```
What's this for? Maybe add a comment.

LarryRuane commented on src/logging.cpp:
```diff
@@ -239,13 +239,86 @@ namespace BCLog {
 
-void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& logging_function, const std::string& source_file, const int source_line)
+static constexpr uint64_t HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION{1024 * 1024};
+
```
Comment needed describing what this function does at a high level and documenting the return value.

dergoegge replied:
There is a comment in `src/logging.h`.

LarryRuane commented on src/logging.cpp:
```diff
@@ -254,6 +327,16 @@ void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& loggi
     str_prefixed = LogTimestampStr(str_prefixed);
 
+    // Rate limit logging to disk to avoid disk filling attacks.
+    bool skip_writing_to_disk_due_to_rate_limiting{false};
+    if (!skip_disk_usage_rate_limiting) {
+        skip_writing_to_disk_due_to_rate_limiting = RateLimit(str_prefixed, logging_function, source_location);
+    }
```
Consider moving this down to just before it's needed.

LarryRuane commented on src/logging.cpp:
```diff
@@ -270,7 +353,7 @@ void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& loggi
         cb(str_prefixed);
     }
-    if (m_print_to_file) {
+    if (m_print_to_file && !skip_writing_to_disk_due_to_rate_limiting) {
```
I'm not sure it's worth creating this variable `skip_writing_to_disk_due_to_rate_limiting`, maybe:
```
if (m_print_to_file && (skip_disk_usage_rate_limiting || !RateLimit(str_prefixed, logging_function, source_location))) {
```

dergoegge replied:
We want the rate limiting messages like "restarting logging from..." to also be printed to the console, so we have to call `RateLimit` before that and store the result. I am still gonna change the name of `skip_writing_to_disk_due_to_rate_limiting` because its a bit long.

jonatack commented on src/init/common.cpp:
```diff
@@ -76,6 +76,7 @@ void AddLoggingArgs(ArgsManager& argsman)
     argsman.AddArg("-logtimemicros", strprintf("Add microsecond precision to debug timestamps (default: %u)", DEFAULT_LOGTIMEMICROS), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-printtoconsole", "Send trace/debug info to console (default: 1 when no -daemon. To disable logging to file, set -nodebuglogfile)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-shrinkdebugfile", "Shrink debug.log file on client startup (default: 1 when no -debug)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-logratelimiting", strprintf("Rate limit debug logging to disk (default: %u)", DEFAULT_LOGRATELIMITING), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
Good point. I think the more specific ones were added later and the idea is to prefer using them for new code.

Thoragh commented on src/logging.cpp:
```diff
@@ -239,13 +239,86 @@ namespace BCLog {
+    // Is the quota exceeded before this log call?
+    bool quota_exceeded_before = quota_usage.m_bytes_logged > HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION;
+
+    bool dont_skip = false;
```
If i understand the code correctly, `dont_skip` is used to make sure the hourly reset messages are printed.
Maybe the code could be simplified by removing `dont_skip` and calling `LogPrintf` instead for the hourly messages?

dergoegge replied:
That is correct!
But this does unfortunately not work because we hold a lock here https://github.com/bitcoin/bitcoin/blob/a2916ae5043436d3fd22acb44c5663227fa7eb90/src/logging.cpp#L317

A recursive call to `LogPrintf` also feels a bit dangerous to me.

dergoegge commented:
[`log_ratelimiting.2`](https://github.com/dergoegge/bitcoin/tree/log_ratelimiting.2) -> [`log_ratelimiting.3`](https://github.com/dergoegge/bitcoin/tree/log_ratelimiting.3)

<details>
<summary>git range-diff log_ratelimiting.2...log_ratelimiting.3</summary>

```c++
1:  01ba3285d ! 1:  2cab018d9 log: Mitigate disk filling attacks by rate limiting LogPrintf
    @@ src/init/common.cpp: void AddLoggingArgs(ArgsManager& argsman)
          argsman.AddArg("-logtimemicros", strprintf("Add microsecond precision to debug timestamps (default: %u)", DEFAULT_LOGTIMEMICROS), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
          argsman.AddArg("-printtoconsole", "Send trace/debug info to console (default: 1 when no -daemon. To disable logging to file, set -nodebuglogfile)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
          argsman.AddArg("-shrinkdebugfile", "Shrink debug.log file on client startup (default: 1 when no -debug)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
    -+    argsman.AddArg("-logratelimiting", strprintf("Rate limit debug logging to disk (default: %u)", DEFAULT_LOGRATELIMITING), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
    ++    argsman.AddArg("-logratelimiting", strprintf("Rate limit debug logging to disk (default: %u)", DEFAULT_LOGRATELIMITING), ArgsManager::ALLOW_BOOL, OptionsCategory::OPTIONS);
      }
      
      void SetLoggingOptions(const ArgsManager& args)
    @@ src/logging.cpp: namespace BCLog {
     +            str = LogTimestampStr(strprintf(
     +                "Not restarting logging from %s:%d (%s): "
     +                "because %d messages (%d MiB) were dropped during the last hour which still exceeds the limit of %d MiB.\n",
    -+                source_location.first, source_location.second, logging_function, quota_usage.m_messages_dropped,
    ++                source_location.m_file, source_location.m_line, logging_function, quota_usage.m_messages_dropped,
     +                quota_usage.m_bytes_dropped / (1024 * 1024), HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION / (1024 * 1024)));
     +        } else if (quota_exceeded_before) {
     +            str = LogTimestampStr(strprintf(
     +                "Restarting logging from %s:%d (%s): "
     +                "%d messages (%d MiB) were dropped during the last hour.\n"
     +                "%s",
    -+                source_location.first, source_location.second, logging_function, quota_usage.m_messages_dropped,
    ++                source_location.m_file, source_location.m_line, logging_function, quota_usage.m_messages_dropped,
     +                quota_usage.m_bytes_dropped / (1024 * 1024), str));
     +            --m_rate_limited_locations;
     +        }
    @@ src/logging.cpp: namespace BCLog {
     +            ">%d MiB logged during the last hour. "
     +            "Suppressing logging to disk from this source location for up to one hour. "
     +            "Console logging unaffected. Last log entry: %s",
    -+            source_location.first, source_location.second, logging_function,
    ++            source_location.m_file, source_location.m_line, logging_function,
     +            HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION / (1024 * 1024), str));
     +        ++m_rate_limited_locations;
     +    } else if (!dont_skip) {
    @@ src/logging.cpp: namespace BCLog {
      
          if (m_log_sourcelocations && m_started_new_line) {
     -        str_prefixed.insert(0, "[" + RemovePrefix(source_file, "./") + ":" + ToString(source_line) + "] [" + logging_function + "] ");
    -+        str_prefixed.insert(0, "[" + RemovePrefix(source_location.first, "./") + ":" + ToString(source_location.second) + "] [" + logging_function + "] ");
    ++        str_prefixed.insert(0, "[" + RemovePrefix(source_location.m_file, "./") + ":" + ToString(source_location.m_line) + "] [" + logging_function + "] ");
          }
      
          if (m_log_threadnames && m_started_new_line) {
    @@ src/logging.cpp: void BCLog::Logger::LogPrintStr(const std::string& str, const s
          str_prefixed = LogTimestampStr(str_prefixed);
      
     +    // Rate limit logging to disk to avoid disk filling attacks.
    -+    bool skip_writing_to_disk_due_to_rate_limiting{false};
    -+    if (!skip_disk_usage_rate_limiting) {
    -+        skip_writing_to_disk_due_to_rate_limiting = RateLimit(str_prefixed, logging_function, source_location);
    -+    }
    ++    bool log_quota_exceeded = !skip_disk_usage_rate_limiting &&
    ++                              RateLimit(str_prefixed, logging_function, source_location);
     +
    ++    // To avoid confusion caused by dropped log messages when debugging an issue, we prefix log lines with "[*]"
    ++    // when there are any supressed source locations.
     +    if (m_rate_limited_locations > 0) {
     +        str_prefixed.insert(0, "[*] ");
     +    }
    @@ src/logging.cpp: void BCLog::Logger::LogPrintStr(const std::string& str, const s
              cb(str_prefixed);
          }
     -    if (m_print_to_file) {
    -+    if (m_print_to_file && !skip_writing_to_disk_due_to_rate_limiting) {
    ++    if (m_print_to_file && !log_quota_exceeded) {
              assert(m_fileout != nullptr);
      
              // reopen the log file, if requested
    @@ src/logging.h: struct LogCategory {
      };
      
     +// Replace with std::source_location when switching to C++20.
    -+using SourceLocation = std::pair<const char*, int>;
    ++struct SourceLocation {
    ++    const char* m_file{nullptr};
    ++    int m_line{0};
    ++
    ++    bool operator==(const SourceLocation& other) const
    ++    {
    ++        return m_file && other.m_file &&
    ++               std::strcmp(m_file, other.m_file) == 0 &&
    ++               m_line == other.m_line;
    ++    }
    ++};
    ++
     +struct SourceLocationHasher {
     +    size_t operator()(const SourceLocation& source_location) const noexcept
     +    {
     +        // Use CSipHasher(0, 0) as a simple way to get uniform distribution.
    -+        return static_cast<size_t>(CSipHasher(0, 0).Write(std::hash<const char*>{}(source_location.first)).Write(std::hash<int>{}(source_location.second)).Finalize());
    ++        return static_cast<size_t>(CSipHasher(0, 0).Write(std::hash<const char*>{}(source_location.m_file)).Write(std::hash<int>{}(source_location.m_line)).Finalize());
     +    }
     +};
     +
    @@ src/logging.h: namespace BCLog {
              std::list<std::string> m_msgs_before_open GUARDED_BY(m_cs);
              bool m_buffering GUARDED_BY(m_cs) = true; //!< Buffer messages before logging can be started.
     +        std::unordered_map<SourceLocation, QuotaUsage, SourceLocationHasher> m_quota_usage_per_source_location GUARDED_BY(m_cs);
    -+        std::atomic_int m_rate_limited_locations;
    ++        int m_rate_limited_locations GUARDED_BY(m_cs) = 0;
      
              /**
               * m_started_new_line is a state variable that will suppress printing of
    @@ src/logging.h: namespace BCLog {
      
     +        /** Evaluate if a log message should not be printed to disk.
     +         *
    -+         * @param[in,out] str                 The log message.
    -+         *                                    The rate limiting logic message may change the message
    -+         *                                    to signal the start or end of a supression period.
    -+         * @param[in]     logging_function    The function name from which the log message originated.
    -+         * @param[in]     source_location     The source location from which the log message originated.
    ++         * @param[in,out] str              The log message.
    ++         *                                 The rate limiting logic message may change the message
    ++         *                                 to signal the start or end of a supression period.
    ++         * @param[in]     logging_function The function name from which the log message originated.
    ++         * @param[in]     source_location  The source location from which the log message originated.
    ++         *
    ++         * @return true if the message should not be printed to disk, false otherwise.
     +         */
     +        bool RateLimit(std::string& str, const std::string& logging_function, const SourceLocation& source_location) EXCLUSIVE_LOCKS_REQUIRED(m_cs);
     +
    @@ src/logging.h: static inline void LogPrintf_(const std::string& logging_function
              }
     -        LogInstance().LogPrintStr(log_msg, logging_function, source_file, source_line);
     +
    -+        const SourceLocation source_location = std::make_pair(source_file, source_line);
    ++        const SourceLocation source_location{source_file, source_line};
     +        LogInstance().LogPrintStr(log_msg, logging_function, source_location, skip_disk_usage_rate_limiting);
          }
      }
2:  a2916ae50 ! 2:  e51bd0bdc test: Add logging test for rate limiting
    @@ src/test/logging_tests.cpp: BOOST_AUTO_TEST_CASE(logging_timer)
          BOOST_CHECK_EQUAL(micro_timer.LogMsg("test micros"), "tests: test micros (1000000.00μs)");
      }
      
    -+void GetLogFileSize(size_t& size)
    ++size_t GetLogFileSize()
     +{
     +    boost::system::error_code ec;
    -+    size = fs::file_size(LogInstance().m_file_path, ec);
    ++    size_t size = fs::file_size(LogInstance().m_file_path, ec);
     +    if (ec) LogPrintf("%s: %s %s\n", __func__, ec.message(), LogInstance().m_file_path);
     +    BOOST_CHECK(!ec);
    ++    return size;
     +}
     +
     +void LogFromFixedLocation(const std::string& str)
    @@ src/test/logging_tests.cpp: BOOST_AUTO_TEST_CASE(logging_timer)
     +    bool prev_log_threadnames = LogInstance().m_log_threadnames;
     +    LogInstance().m_log_threadnames = false;
     +
    ++    // Log 1024-character lines (1023 plus newline) to make the math simple.
     +    std::string log_message(1023, 'a');
     +
     +    SetMockTime(std::chrono::seconds{1});
     +
     +    size_t prev_log_file_size, curr_log_file_size;
    -+    GetLogFileSize(prev_log_file_size);
    ++    prev_log_file_size = GetLogFileSize();
     +
     +    // Log 1 MiB, this should be allowed.
     +    for (int i = 0; i < 1024; ++i) {
     +        LogFromFixedLocation(log_message);
     +    }
    -+    GetLogFileSize(curr_log_file_size);
    ++    curr_log_file_size = GetLogFileSize();
     +    BOOST_CHECK(curr_log_file_size - prev_log_file_size == 1024 * 1024);
     +
     +    LogFromFixedLocation("This should trigger rate limiting");
    -+    GetLogFileSize(prev_log_file_size);
    ++    prev_log_file_size = GetLogFileSize();
     +
     +    // Log 0.5 MiB, this should not be allowed and all messages should be dropped.
     +    for (int i = 0; i < 512; ++i) {
     +        LogFromFixedLocation(log_message);
     +    }
    -+    GetLogFileSize(curr_log_file_size);
    ++    curr_log_file_size = GetLogFileSize();
     +    BOOST_CHECK(curr_log_file_size - prev_log_file_size == 0);
     +
     +    // Let one hour pass.
     +    SetMockTime(std::chrono::seconds{60 * 60 + 1});
     +    LogFromFixedLocation("This should trigger the quota usage reset");
    -+    GetLogFileSize(prev_log_file_size);
    ++    prev_log_file_size = GetLogFileSize();
     +
     +    // Log 1 MiB, this should be allowed since the usage was reset.
     +    for (int i = 0; i < 1024; ++i) {
     +        LogFromFixedLocation(log_message);
     +    }
    -+    GetLogFileSize(curr_log_file_size);
    ++    curr_log_file_size = GetLogFileSize();
     +    BOOST_CHECK(curr_log_file_size - prev_log_file_size == 1024 * 1024);
     +
     +    LogFromFixedLocation("This should trigger rate limiting");
    -+    GetLogFileSize(prev_log_file_size);
    ++    prev_log_file_size = GetLogFileSize();
     +
     +    // Log 1 MiB, this should not be allowed and all messages should be dropped.
     +    for (int i = 0; i < 1024; ++i) {
     +        LogFromFixedLocation(log_message);
     +    }
    -+    GetLogFileSize(curr_log_file_size);
    ++    curr_log_file_size = GetLogFileSize();
     +    BOOST_CHECK(curr_log_file_size - prev_log_file_size == 0);
     +
     +    LogFromFixedLocation("This should also be dropped and bring the total of dropped bytes above 1 MiB.");
    -+    GetLogFileSize(curr_log_file_size);
    ++    curr_log_file_size = GetLogFileSize();
     +    BOOST_CHECK(curr_log_file_size - prev_log_file_size == 0);
     +
     +    // Let another hour pass
     +    SetMockTime(std::chrono::seconds{2 * (60 * 60 + 1)});
     +    LogFromFixedLocation("Normally this would reset the usage but it does not because we dropped more than 1 MiB in the previous hour.");
    -+    GetLogFileSize(prev_log_file_size);
    ++    prev_log_file_size = GetLogFileSize();
     +
     +    // Log 1 MiB, this should not be allowed and all messages should be dropped.
     +    for (int i = 0; i < 1024; ++i) {
     +        LogFromFixedLocation(log_message);
     +    }
    -+    GetLogFileSize(curr_log_file_size);
    ++    curr_log_file_size = GetLogFileSize();
     +    BOOST_CHECK(curr_log_file_size - prev_log_file_size == 0);
     +
     +    // Let another hour pass
     +    SetMockTime(std::chrono::seconds{3 * (60 * 60 + 1)});
     +    LogFromFixedLocation("This should trigger the quota usage reset");
    -+    GetLogFileSize(prev_log_file_size);
    ++    prev_log_file_size = GetLogFileSize();
     +
     +    // Log 1 MiB, this should be allowed since the usage was reset.
     +    for (int i = 0; i < 1024; ++i) {
     +        LogFromFixedLocation(log_message);
     +    }
    -+    GetLogFileSize(curr_log_file_size);
    ++    curr_log_file_size = GetLogFileSize();
     +    BOOST_CHECK(curr_log_file_size - prev_log_file_size == 1024 * 1024);
     +
     +    LogInstance().m_log_timestamps = prev_log_timestamps;

```

</details>

LarryRuane commented:
Reviewers, please take a look at this alternate approach: https://github.com/bitcoin/bitcoin/pull/22350.

narula commented:
fwiw prefer the more traditional log rotation approach in #22350; i find it cognitively easier to think in terms of stale logs rather than source suppression.

practicalswift commented:
@dergoegge Would you mind rebasing? I would like to re-review it after rebase :)

FWIW I prefer this solution to the more complex log rotation approach taken in #22350. I find the simpler approach taken in this PR much easier to reason about. This PR solves the issue at hand in the simplest possible way IMO.

dergoegge commented:
@practicalswift Thanks for the nudge! :)

Rebased on latest master.

dergoegge committed:
[log] Introduce log rate limiter class

The LogRatelimiter class implements a fixed window rate limiter. The
rate limiter allows a fixed amount of bytes to be consumed within a
fixed time window.

dergoegge committed:
[log] Introduce source location type

The SourceLocation type stores the filename and line of a source code
location.
In a later commit we use this type as the key type in an unordered map
and set to keep track of rate limters for each location.

dergoegge committed:
[test util] Mark ~DebugLogHelper as noexcept(false)

We mark ~DebugLogHelper as noexcept(false) to be able to catch the
exception it throws. This lets us use it in test in combination with
BOOST_CHECK_THROW and BOOST_CHECK_NO_THROW to check that certain log
messages are (not) logged.

dergoegge committed:
[log] Add rate limiting to LogPrintf

To mitigate disk filling attacks caused by unsafe usages of LogPrintf,
we rate limit LogPrintf by using the fixed window rate limiter
(BCLog::LogRatelimiter) introduced in an earlier commit.

The rate limiting logic is applied per source location instead of
globally. A source location is allowed to log up to 1 MiB per hour.
Source locations that violate the limit will have their logs supressed
for up to one hour.

dergoegge committed:
[config] Add -ratelimitlogging config option

The -ratelimitlogging can be used to enable/disable the rate limiting to
disk. Rate limiting is enabled by default.

dergoegge committed:
[validation] Exempt UpdateTipLog from rate limiting

UpdateTipLog logs everytime a new tip is activated. This occurs at an
increased frequency during IBD and should therefore be exempt from rate
limiting.

dergoegge committed:
[test] Check for expected log rate limiting messages

dergoegge committed:
[test] Test for expected file size changes when rate limiting is enabled

dergoegge committed:
[test] Check that log rate limiting is disabled for exempt source locations

dergoegge committed:
[test] Check that rate limiting can be disabled

dergoegge commented:
In an attempt to make this more reviewable, i split the changes into multiple commits.

@practicalswift pinging you, since you were previously interested in reviewing

jnewbery commented on src/test/logging_tests.cpp:
```diff
@@ -30,4 +31,140 @@ BOOST_AUTO_TEST_CASE(logging_timer)
+    case 3:
+        LogPrint(BCLog::ALL, "%s\n", message);
+        break;
+    default:
+        BOOST_CHECK_MESSAGE(false, "Unkown log location");
```
This shouldn't be a `BOOST_CHECK`, which is used to check program logic. This is to catch errors in the unit test code. I think just remove it and let the compiler warn.

```suggestion
    // no default case, so the compiler can warn about missing cases
```

jnewbery commented on src/logging.h:
```diff
@@ -181,18 +257,31 @@ static inline void LogPrintf_(const std::string& logging_function, const std::st
+// MUST NOT be rate limited no matter how often they are logged. That requirement
+// should be extremely rare, so please use with care. Prefer LogPrintf(...) if
+// possible.
+#define LogPrintfWithoutRateLimiting(...) LogPrintf_(__func__, __FILE__, __LINE__, /* skip_ratelimiting */ true, __VA_ARGS__)
```
I think it'd be nice to unify the various logprint macros by introducing new categories:

- `UNCONDITIONAL_RATE_LIMITED` replaces `LogPrintf`
- `UNCONDITIONAL_ALWAYS` replaces `LogPrintfWithoutRateLimiting`

or similar.

dergoegge replied:
Are you suggesting to get rid of `LogPrintf` completely or to define it as the following?
```c++
#define LogPrintf(...) LogPrint(UNCONDITIONAL_RATE_LIMITED, __VA_ARGS__)
```
I would not be a fan of replacing every usage of `LogPrintf` in this PR.

jnewbery replied:
I think it'd be good to get rid of `LogPrintf` completely in the future, but I'm not suggesting you do that in this PR. I'd just prefer that his PR doesn't add a new log macro.

jnewbery commented on src/logging.cpp:
```diff
@@ -247,13 +249,13 @@ namespace BCLog {
 } // namespace BCLog
 
-void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& logging_function, const std::string& source_file, const int source_line)
+void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& logging_function, const SourceLocation& source_location, bool skip_ratelimiting)
```
Do we have any idea if the new logic in this function impacts performance?

dergoegge replied:
Here are some benchmarks using the code from #18815:

merge base:
```
|               ns/op |                op/s |    err% |          ins/op |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|---------------:|--------:|----------:|:----------
|                3.63 |      275,623,892.70 |    0.5% |           54.00 |          13.00 |    0.0% |      0.01 | `LoggingNoCategory`
|               40.65 |       24,600,010.03 |    0.9% |          659.00 |         155.00 |    0.0% |      0.01 | `LoggingNoFile`
|            2,165.69 |          461,746.80 |    0.2% |       19,809.05 |       3,709.01 |    0.1% |      0.01 | `LoggingNoThreadNames`
|            2,189.45 |          456,736.09 |    0.8% |       19,842.05 |       3,717.01 |    0.2% |      0.01 | `LoggingYoCategory`
|            2,224.15 |          449,609.33 |    0.9% |       20,373.05 |       3,839.01 |    0.1% |      0.01 | `LoggingYoThreadNames`

```

pull request:
```
|               ns/op |                op/s |    err% |          ins/op |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|---------------:|--------:|----------:|:----------
|                3.88 |      257,848,736.66 |    1.2% |           54.00 |          13.00 |    0.0% |      0.01 | `LoggingNoCategory`
|               40.29 |       24,820,332.51 |    2.5% |          663.00 |         155.00 |    0.0% |      0.01 | `LoggingNoFile`
|            2,440.31 |          409,783.92 |    1.5% |       22,024.06 |       4,017.01 |    0.1% |      0.01 | `LoggingNoThreadNames`
|            2,486.87 |          402,111.61 |    4.6% |       21,513.06 |       3,951.01 |    0.1% |      0.01 | `LoggingYoCategory`
|            2,571.68 |          388,850.52 |    3.3% |       22,584.06 |       4,147.01 |    0.1% |      0.01 | `LoggingYoThreadNames`
```

I would say that the additional 200 - 300 nano seconds per `LogPrintf` call are negligible.

dergoegge commented on src/test/logging_tests.cpp:
```diff
@@ -30,4 +31,140 @@ BOOST_AUTO_TEST_CASE(logging_timer)
+
+    SetMockTime(std::chrono::hours{1});
+
+    size_t log_file_size = std::filesystem::file_size(LogInstance().m_file_path);
```
I think i would like to keep them, as `LogFromLocationAndExpect` uses the log callbacks which fire regardless of whether or not the message is actually printed to the console or disk.

I think these checks make sense just to be really sure that the file sizes change as expected. I have had bugs while writing this where the messages from the rate limiting logic where printed at the wrong time, so simply relying on those is not enough IMO.

dergoegge committed:
[log] Add two new categories for unconditional logging

We create two new categories `UNCONDITIONAL_ALWAYS` and
`UNCONDITIONAL_RATE_LIMITED` that are always enabled by default.

LogPrintf now logs using the `UNCONDITIONAL_RATE_LIMITED` category which
will start to apply rate limiting in a later commit.

For some log locations it might be safe to allow more frequent logging
without rate limiting. These locations should use the
`UNCONDITIONAL_ALWAYS` category.

dergoegge commented on src/logging.cpp:
```diff
@@ -262,7 +264,43 @@ void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& loggi
+        m_ratelimiters.emplace(source_location, LogRatelimiter{RATELIMIT_WINDOW_SIZE, RATELIMIT_WINDOW_MAX_BYTES});
+    }
+
+    uint64_t dropped_bytes = m_ratelimiters[source_location].GetDroppedBytes();
```
I dropped the non-default constructor since we only ever constructed with `RATELIMIT_WINDOW_SIZE` and `RATELIMIT_WINDOW_MAX_BYTES`. So now a new `LogRateLimiter` will be constructed using the default constructor when `[]` is used the first time.

dergoegge commented on src/logging.h:
```diff
@@ -66,6 +94,49 @@ namespace BCLog {
+            : m_reset_interval{interval}, m_available_bytes{max_bytes}, m_max_bytes{max_bytes}
+        {
+            m_last_reset = GetTime<std::chrono::seconds>();
+        }
```
See https://github.com/bitcoin/bitcoin/pull/21603#discussion_r824828452

glozow commented on src/logging.cpp:
```diff
@@ -165,6 +164,8 @@ const CLogCategoryDesc LogCategories[] =
 #endif
     {BCLog::UTIL, "util"},
     {BCLog::BLOCKSTORE, "blockstorage"},
+    {BCLog::UNCONDITIONAL_RATE_LIMITED, "uncond_rate_limited"},
+    {BCLog::UNCONDITIONAL_ALWAYS, "uncond_always"},
```
Just to clarify - all the other categories (`QT`, `REINDEX`, `BENCH`, etc.) are not rate-limited?

dergoegge replied:
Yes, only uses of `LogPrintf` (`LogPrint(BCLog::UNCONDITIONAL_RATE_LIMITED, ...)`) are rate limited.

glozow reviewed (commented):
Concept ACK 

Re: source-based vs global rate-limiting or rotation: from what I understand, the disk-filling attack we're trying to prevent is an attacker figuring out some unconditional logging they can trigger over and over again, so a source-based rate-limiting approach seems to be the correct solution. I assume there aren't other attacks we're looking to prevent?

dergoegge commented on src/init/common.cpp:
```diff
@@ -76,6 +76,7 @@ void AddLoggingArgs(ArgsManager& argsman)
     argsman.AddArg("-logtimemicros", strprintf("Add microsecond precision to debug timestamps (default: %u)", DEFAULT_LOGTIMEMICROS), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-printtoconsole", "Send trace/debug info to console (default: 1 when no -daemon. To disable logging to file, set -nodebuglogfile)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-shrinkdebugfile", "Shrink debug.log file on client startup (default: 1 when no -debug)", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-ratelimitlogging", strprintf("Rate limit debug logging to disk (default: %u)", DEFAULT_RATELIMITLOGGING), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
```
Replaced "debug" with "unconditional".

dergoegge commented:
Thanks for the review @mzumsande @glozow!

> [...] the disk-filling attack we're trying to prevent is an attacker figuring out some unconditional logging they can trigger over and over again [...]
> I assume there aren't other attacks we're looking to prevent?

None that I can think of, this PR is aimed at exactly what you described.

jnewbery commented on src/logging.h:
```diff
@@ -85,7 +158,7 @@ namespace BCLog {
 
         /** Log categories bitfield. */
-        std::atomic<uint32_t> m_categories{0};
+        std::atomic<uint32_t> m_categories{BCLog::UNCONDITIONAL_RATE_LIMITED | BCLog::UNCONDITIONAL_ALWAYS};
```
Does setting the default to `BCLog::UNCONDITIONAL_RATE_LIMITED | BCLog::UNCONDITIONAL_ALWAYS` break the logic in `DefaultShrinkDebugFile()`?

dergoegge replied:
Good catch, fixed!

jnewbery commented on src/logging.cpp:
```diff
@@ -337,3 +374,26 @@ void BCLog::Logger::ShrinkDebugFile()
+    }
+}
+
+bool BCLog::LogRateLimiter::Consume(uint64_t bytes)
```
Is it possible for this to return false followed by true, even if the rate limiter doesn't reset, eg:

- m_available_bytes is 100
- `Consume(110)` is called and returns false
- `Consume(90)` is called and returns true.

Which would then cause the "Restarting logging from ..." and "Excessive logging detected from..." logs to be called before the LogRateLimiter has been reset.

If so, I think this is the fix:

```diff
    if (bytes > m_available_bytes) {
        m_dropped_bytes += bytes;
+       m_available_bytes = 0;
        return false;
    }

dergoegge replied:
Yes, trying to consume more bytes than are available should result in no available bytes afterwards.

jnewbery commented on src/logging.cpp:
```diff
@@ -264,7 +265,43 @@ void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& loggi
+
+        // To avoid confusion caused by dropped log messages when debugging an issue,
+        // we prefix log lines with "[*]" when there are any supressed source locations.
+        if (m_supressed_locations.size() > 0) {
+            str_prefixed.insert(0, "[*] ");
+        }
```
Should this be outside the `skip_rate_limiting` conditional, so it is prefixed to all lines?

```diff
                 source_location.m_file, source_location.m_line, logging_function,
                 LogRateLimiter::WINDOW_MAX_BYTES / (1024 * 1024), str_prefixed));
         }
+    }
 
-        // To avoid confusion caused by dropped log messages when debugging an issue,
-        // we prefix log lines with "[*]" when there are any supressed source locations.
-        if (m_supressed_locations.size() > 0) {
-            str_prefixed.insert(0, "[*] ");
-        }
+    // To avoid confusion caused by dropped log messages when debugging an issue,
+    // we prefix log lines with "[*]" when there are any supressed source locations.
+    if (m_supressed_locations.size() > 0) {
+        str_prefixed.insert(0, "[*] ");
     }
```

jnewbery commented on src/logging.cpp:
```diff
@@ -334,19 +339,20 @@ namespace BCLog {
     std::string str_prefixed = LogEscapeMessage(str);
 
-    if ((category != LogFlags::NONE || level != Level::None) && m_started_new_line) {
+    const bool print_category{category != LogFlags::NONE && category != LogFlags::UNCONDITIONAL_ALWAYS && category != LogFlags::UNCONDITIONAL_RATE_LIMITED};
```
Is the `category != LogFlags::NONE` necessary here? I don't think anything logs with category NONE now.

dergoegge replied:
There are still tests that log using NONE:
https://github.com/bitcoin/bitcoin/blob/a6377c5fe4f8d8fabbae527df888afdd9c6f771a/src/test/logging_tests.cpp#L80-L81

What do you think of removing the NONE category entirely?

jnewbery replied:
I think those should be replaced with `LogPrintf_(..., BCLog::LogFlags::UNCONDITIONAL_ALWAYS, ...)`, since it's impossible for `LogPrintf_(..., BCLog::LogFlags::NONE, ...)` to be called in the produce code.

It seems fine to remove `BCLog::LogFlags::NONE` after that, but I don't think it's necessary in this PR.

jnewbery commented:
utACK a6377c5fe4f8d8fabbae527df888afdd9c6f771a

MarcoFalke commented on src/logging.cpp:
```diff
@@ -7,7 +7,6 @@
 #include <logging.h>
 #include <util/threadnames.h>
 #include <util/string.h>
-#include <util/time.h>
```
Why are you removing this include when it is needed?

MarcoFalke replied:
According to the dev notes and iwyu it will need to be included in both files if both files use it.

MarcoFalke commented on src/logging.h:
```diff
@@ -75,6 +105,47 @@ namespace BCLog {
+
+        LogRateLimiter()
+            : m_last_reset{GetTime<std::chrono::seconds>()},
+              m_available_bytes{WINDOW_MAX_BYTES} {}
```
Could drop this and use C++11 member init?

jonatack commented:
>     * Only `LogPrintf(…)` is rate limited. `LogPrint(category, …)` (`-debug`) is not rate limited.

It looks like many of the LogPrintf messages will be going away with the update to severity-based logging and the added complexity here may be worth reevaluating (and potentially avoiding) after that.

MarcoFalke commented on src/logging.h:
```diff
@@ -75,6 +105,47 @@ namespace BCLog {
+    {
+    private:
+        //! Timestamp of the last window reset.
+        std::chrono::seconds m_last_reset;
```
```suggestion
        std::chrono::steady_clock::time_point m_last_reset{std::chrono::steady_clock::now()};
```

Looks like you are not mocking this in the tests, so could use steady clock?

If not, new code should probably use `NodeClock::now()` over `GetTime()`.

dergoegge replied:
Changed it to use `NodeClock`.

maflcko reviewed (commented):
left some nits (feel free to ignore for now)

dergoegge commented:
> It looks like many of the LogPrintf messages will be going away with the update to severity-based logging and the added complexity here may be worth reevaluating (and potentially avoiding) after that.

@jonatack IMO removing some of the `LogPrintf` locations (or assigning them a severity level which is not logged by default) does not really address the issue at its root like this PR is trying to do. Any remaining default logging could still cause issues as well as any logging that is introduced in the future.

jonatack commented:
What I am working on is that no logging on by default would be externally provokable (and potentially removing LogPrintf).

mzumsande commented:
> What I am working on is that no logging on by default would be externally provokable (and potentially removing LogPrintf). That along with reducing default logging in general has appeared to be the desired direction for some time now AFAICS. Of course, this could still be merged in the interim and then removed if no longer needed, but I'm targeting the next release.

With https://github.com/bitcoin/bitcoin/pull/25202/commits/bd971bffb02c7b06aac9a479f7e5ed8f71de2bec, category-based logging with a severity `>= BCLog::Level::Warning` is equivalent to `LogPrintf` from the viewpoint of disk filling attacks.
I think that it would be really good to have this general countermeasure in place, especially with probable future PRs that convert more existing log messages to the new format, because we'll now have to be careful about not adding category-based warnings or errors that could be triggered externally. The old rule, which is no longer valid but probably still in many contributors' heads was that one only needs to be careful about `LogPrintf`, while category-base logging is not so critical.

jonatack commented:
> convert more existing log messages to the new format, because we'll now have to be careful about not adding category-based warnings or errors that could be triggered externally

Yes, doing precisely this right now for all of them.

MarcoFalke commented:
Is it possible to convert *all* of them? I think currently we don't have a way to express a log category but unconditional log severity. Unless you want to use `Warning`/`Error` for "unconditional".

jonatack commented:
@MarcoFalke added that in #25203.

jonatack commented:
> Good point, I think that this would definitely be useful (here or as a follow-up) - having a rate limit only for `LogPrintf` but not for unconditionally logged category-based warnings and errors would be confusing.

I agree; though just updated the (copious) net processing logging and there are a only a few remaining unconditional messages:
```
src/net_processing.cpp:2305:                LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "Large reorg, won't direct fetch to %s (%d)\n",
src/net_processing.cpp:2702:            LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "connected to self at %s, disconnecting\n", pfrom.addr.ToString());
src/net_processing.cpp:3411:                    LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "Not relaying non-mempool transaction %s from forcerelay peer=%d\n", tx.GetHash().ToString(), pfrom.GetId());
src/net_processing.cpp:4176:        LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "Not punishing misbehaving peer %d because it has noban permission set!\n", peer.m_id);
src/net_processing.cpp:4182:        LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "Not punishing misbehaving peer %d because it was manually connected!\n", peer.m_id);
src/net_processing.cpp:4189:        LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "Warning: disconnecting but not discouraging %s local peer %d!\n",
```

dergoegge commented:
> having a rate limit only for LogPrintf but not for unconditionally logged category-based warnings and errors would be confusing.

Also agree with this and i want to add that to this PR.

The two new categories I introduce here (`UNCONDITIONAL_ALWAYS` and `UNCONDITIONAL_RATE_LIMITED`) don't really make sense anymore, given that we now also have category based unconditional logging. Whether a log location should be rate limited could purely be based on the severity level.
I was thinking of the following: Adding an extra severity `Always` which is logged unconditionally without rate limiting (e.g. for [this](https://github.com/bitcoin/bitcoin/blob/1b2e1d179c5b350cac69ee670e01355314f25e11/src/validation.cpp#L2517) location) and rate limiting logging to disk if `severity == None || (severity >= Warning && severity < Always)`. Additionally, categories specified by `-debug=cat` should still be logged unconditionally without rate limiting.  @mzumsande @jonatack @MarcoFalke what do y'all think of that?

MarcoFalke commented:
If there is a log category with `Info`, which is also the default level, you wouldn't need an `Always` level?

mzumsande commented:
> If there is a log category with `Info`, which is also the default level, you wouldn't need an `Always` level?

I don't see how - the "Always" level is meant for noisy unconditional logs that can hit the rate limit, but are so important that they should be exempt from it - basically only `UpdateTipLog` during IBD.

jonatack commented:
The logging you might be worried about limiting would be, with the current direction, IIRC:

- LogPrintf
- LogPrintfWithCategory
- LogPrintLevel with level Info/Warning/Error

dergoegge commented:
Haven't had the time to maintain this. I still think some general mitigation for disk-filling would be nice, so I might pick this up again later unless someone else beats me to it. Closing for now.
